src/common/meta/src/ddl/drop_database/metadata.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_procedure::Status;
+use serde::{Deserialize, Serialize};
+
+use super::end::DropDatabaseEnd;
+use crate::ddl::drop_database::{DropDatabaseContext, State};
+use crate::ddl::DdlContext;
+use crate::error::Result;
+use crate::key::schema_name::SchemaNameKey;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct DropDatabaseRemoveMetadata;
+
+#[async_trait::async_trait]
+#[typetag::serde]
+impl State for DropDatabaseRemoveMetadata {
+    async fn next(
+        &mut self,
+        ddl_ctx: &DdlContext,
+        ctx: &mut DropDatabaseContext,
+    ) -> Result<(Box<dyn State>, Status)> {
+        ddl_ctx
+            .table_metadata_manager
+            .schema_manager()
+            .delete(SchemaNameKey::new(&ctx.catalog, &ctx.schema))
```

Comment:
Yes. Firstly, the procedure locks can ensure no other tables were created in the schema(same as the schema). Secondly, the unfinished procedures will be recovered(enqueued) before any future procedures are accepted (related to #3549).

Reply:
What if the drop database procedure failed not because of the leader switching, but within itself? For example, some requests to datanodes are timeout and beyond retries.

Reply:
> What if the drop database procedure failed not because of the leader switching, but within itself? For example, some requests to datanodes are timeout and beyond retries.

Every step should be idempotent, and I will add some tests to ensure this point.

Reply:
I mean if the drop database procedure is failed here, but the metasrv is good. During this period, a new table could be created under the database. When the drop database procedure is running again, it starts from here - deleting the database metadata directly.

Reply:
During the retrying, the failed procedure won't release the locks; therefore, the table creation will be executed after the failed procedure exists